File: backend/app/joysafeter_orchestrator/runtime_config.py
"""Runtime-tunable configuration with thread-safe reads.

Ported from joysafeter-types/src/runtime_config.rs.
Values can be updated without restarting the process.
"""
import threading
from typing import Optional


class RuntimeConfig:
    def __init__(
        self,
        idle_timeout_sec: int = 600,
        stopped_max_age_sec: int = 3600,
        heartbeat_timeout_sec: int = 30,
        sandbox_failure_threshold: int = 3,
        pool_min_size: int = 2,
        pool_max_age_sec: int = 1800,
        event_batch_max_size: int = 50,
        event_batch_max_delay_ms: int = 50,
    ):
        self._lock = threading.Lock()
        self._idle_timeout_sec = idle_timeout_sec
        self._stopped_max_age_sec = stopped_max_age_sec
        self._heartbeat_timeout_sec = heartbeat_timeout_sec
        self._sandbox_failure_threshold = sandbox_failure_threshold
        self._pool_min_size = pool_min_size
        self._pool_max_age_sec = pool_max_age_sec
        self._event_batch_max_size = event_batch_max_size
        self._event_batch_max_delay_ms = event_batch_max_delay_ms

    @property
    def idle_timeout_sec(self) -> int:
        with self._lock:
            return self._idle_timeout_sec

    @idle_timeout_sec.setter
    def idle_timeout_sec(self, value: int) -> None:
        with self._lock:
            self._idle_timeout_sec = value

    @property
    def stopped_max_age_sec(self) -> int:
        with self._lock:
            return self._stopped_max_age_sec

    @stopped_max_age_sec.setter
    def stopped_max_age_sec(self, value: int) -> None:
        with self._lock:
            self._stopped_max_age_sec = value

    @property
    def heartbeat_timeout_sec(self) -> int:
        with self._lock:
            return self._heartbeat_timeout_sec

    @heartbeat_timeout_sec.setter
    def heartbeat_timeout_sec(self, value: int) -> None:
        with self._lock:
            self._heartbeat_timeout_sec = value

    @property
    def sandbox_failure_threshold(self) -> int:
        with self._lock:
            return self._sandbox_failure_threshold

    @sandbox_failure_threshold.setter
    def sandbox_failure_threshold(self, value: int) -> None:
        with self._lock:
            self._sandbox_failure_threshold = value

    @property
    def pool_min_size(self) -> int:
        with self._lock:
            return self._pool_min_size

    @pool_min_size.setter
    def pool_min_size(self, value: int) -> None:
        with self._lock:
            self._pool_min_size = value

    @property
    def pool_max_age_sec(self) -> int:
        with self._lock:
            return self._pool_max_age_sec

    @pool_max_age_sec.setter
    def pool_max_age_sec(self, value: int) -> None:
        with self._lock:
            self._pool_max_age_sec = value

    @property
    def event_batch_max_size(self) -> int:
        with self._lock:
            return self._event_batch_max_size

    @event_batch_max_size.setter
    def event_batch_max_size(self, value: int) -> None:
        with self._lock:
            self._event_batch_max_size = value

    @property
    def event_batch_max_delay_ms(self) -> int:
        with self._lock:
            return self._event_batch_max_delay_ms

    @event_batch_max_delay_ms.setter
    def event_batch_max_delay_ms(self, value: int) -> None:
        with self._lock:
            self._event_batch_max_delay_ms = value

    def update(self, **kwargs) -> None:
        with self._lock:
            for key, value in kwargs.items():
                attr = f"_{key}"
                if hasattr(self, attr):
                    setattr(self, attr, value)

File: backend/app/joysafeter_orchestrator_rs/src/runtime_config.rs
use std::sync::atomic::{AtomicU32, AtomicU64, Ordering};

/// Runtime-tunable configuration with atomic reads.
///
/// Mirrors the Python `RuntimeConfig`. Values can be updated without
/// restarting the process (via SIGHUP handler).
pub struct RuntimeConfig {
    idle_timeout_sec: AtomicU64,
    stopped_max_age_sec: AtomicU64,
    heartbeat_timeout_sec: AtomicU64,
    sandbox_failure_threshold: AtomicU32,
    pool_min_size: AtomicU64,
    pool_max_age_sec: AtomicU64,
    event_batch_max_size: AtomicU64,
    event_batch_max_delay_ms: AtomicU64,
}

impl RuntimeConfig {
    pub fn new(
        idle_timeout_sec: u64,
        stopped_max_age_sec: u64,
        heartbeat_timeout_sec: u64,
        sandbox_failure_threshold: u32,
        pool_min_size: u64,
        pool_max_age_sec: u64,
        event_batch_max_size: u64,
        event_batch_max_delay_ms: u64,
    ) -> Self {
        Self {
            idle_timeout_sec: AtomicU64::new(idle_timeout_sec),
            stopped_max_age_sec: AtomicU64::new(stopped_max_age_sec),
            heartbeat_timeout_sec: AtomicU64::new(heartbeat_timeout_sec),
            sandbox_failure_threshold: AtomicU32::new(sandbox_failure_threshold),
            pool_min_size: AtomicU64::new(pool_min_size),
            pool_max_age_sec: AtomicU64::new(pool_max_age_sec),
            event_batch_max_size: AtomicU64::new(event_batch_max_size),
            event_batch_max_delay_ms: AtomicU64::new(event_batch_max_delay_ms),
        }
    }

    pub fn from_config(config: &crate::config::JoySafeterConfig) -> Self {
        Self::new(
            config.sandbox_idle_timeout,
            config.sandbox_stopped_ttl,
            config.heartbeat_ttl,
            config.sandbox_failure_threshold,
            config.sandbox_pool_min_size as u64,
            config.sandbox_pool_max_age,
            config.event_batch_max_size as u64,
            config.event_batch_max_delay_ms,
        )
    }

    // Getters
    pub fn idle_timeout_sec(&self) -> u64 {
        self.idle_timeout_sec.load(Ordering::Relaxed)
    }
    pub fn stopped_max_age_sec(&self) -> u64 {
        self.stopped_max_age_sec.load(Ordering::Relaxed)
    }
    pub fn heartbeat_timeout_sec(&self) -> u64 {
        self.heartbeat_timeout_sec.load(Ordering::Relaxed)
    }
    pub fn sandbox_failure_threshold(&self) -> u32 {
        self.sandbox_failure_threshold.load(Ordering::Relaxed)
    }
    pub fn pool_min_size(&self) -> u64 {
        self.pool_min_size.load(Ordering::Relaxed)
    }
    pub fn pool_max_age_sec(&self) -> u64 {
        self.pool_max_age_sec.load(Ordering::Relaxed)
    }
    pub fn event_batch_max_size(&self) -> u64 {
        self.event_batch_max_size.load(Ordering::Relaxed)
    }
    pub fn event_batch_max_delay_ms(&self) -> u64 {
        self.event_batch_max_delay_ms.load(Ordering::Relaxed)
    }

    /// Update all values atomically (called from SIGHUP handler).
    pub fn update(
        &self,
        idle_timeout_sec: u64,
        stopped_max_age_sec: u64,
        heartbeat_timeout_sec: u64,
        sandbox_failure_threshold: u32,
        pool_min_size: u64,
        pool_max_age_sec: u64,
        event_batch_max_size: u64,
        event_batch_max_delay_ms: u64,
    ) {
        self.idle_timeout_sec
            .store(idle_timeout_sec, Ordering::Relaxed);
        self.stopped_max_age_sec
            .store(stopped_max_age_sec, Ordering::Relaxed);
        self.heartbeat_timeout_sec
            .store(heartbeat_timeout_sec, Ordering::Relaxed);
        self.sandbox_failure_threshold
            .store(sandbox_failure_threshold, Ordering::Relaxed);
        self.pool_min_size.store(pool_min_size, Ordering::Relaxed);
        self.pool_max_age_sec
            .store(pool_max_age_sec, Ordering::Relaxed);
        self.event_batch_max_size
            .store(event_batch_max_size, Ordering::Relaxed);
        self.event_batch_max_delay_ms
            .store(event_batch_max_delay_ms, Ordering::Relaxed);
    }
}
